Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+
+## Design
+
+The design proposal is to allow third parties to create their own custom remoting connections by deriving from and providing custom implementations of the `RunspaceConnectionInfo` and `ClientSessionTransportManagerBase` abstract classes.
+These implementations would be part of a PowerShell binary module, which also includes a cmdlet that returns a live connection instance as a `PSSession` object.
+The `PSSession` object can then be passed to existing PowerShell remoting cmdlets to interact with the session.
+The implementation binary can also include other relevant cmdlets and expose public APIs for creating and managing the custom connection.
+
+### RunspaceConnectionInfo
+
+The `System.Management.Automation.Runspaces.RunspaceConnectionInfo` abstract class is already publicly accessible.
+Derivations of this class are used to define specific connection types, including new properties needed for the connection or authentication.
+A derivation must also provide an implementation for the `CreateClientSessionTransportManager` method, which returns an instance of the client session transport manager for the new connection.  
+
+### ClientSessionTransport
+
+The `ClientSessionTransportManagerBase` abstract class is not currently publicly accessible, and will need to be made public.
+Derivations of this class are used to bind live connection data streams to the PowerShell remoting layer.  
+
+Once a connection is established, the data streams are bound to internal PowerShell remoting components through a `ClientSessionTransportManager`, which is derived from the public `ClientSessionTransportManagerBase` abstract class.  
+
+The derived class will override a few virtual functions, the most important of which is the `CreateAsync` method.
+This method runs immediately before protocol messaging begins, and needs to perform two vital functions:
+
+- Create a text writer wrapper object, that is used to write protocol messages to the target input data stream.
+
+- Start a listening thread for protocol messages from the target output data stream.
+
+### Data Streams
+
+All connection data streams used in the transport must be derived from .NET TextReader or TextWriter classes.
+Connections typically have two data streams, one for writing messages to the target, and one for reading messages from the target.
+However, some connections have a third error stream that require a special reader thread.
+An example would be a Process connection that uses stdOut for reading from the target, stdIn for writing to the client, and stdError for reading connection error messages.
+The stdError messages are not part of the remoting protocol and need to be handled separately.
+
+### Creating a connection
+
+The PowerShell module implementing the new remote connection should supply all necessary cmdlets for creating and managing the connection.
+In particular it should have a cmdlet that returns a `PSSession` object.
```

Comment:
We would have to make [PSSession](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.runspaces.pssession?view=powershellsdk-7.0.0) constructor public. Currently it's [internal](https://github.com/PowerShell/PowerShell/blob/658837323599ab1c7a81fe66fcd43f7420e4402b/src/System.Management.Automation/engine/remoting/client/remoterunspaceinfo.cs#L249) stopping someone from creating an instance of this without using reflection. I'm not sure if that should be mentioned under the other 2 classes called out.

Reply:
Yes.  I feel many of the implementation details are trivial and don't need to be mentioned in this proposal.

Reply:
Just for the sake of the discussion, +1 on making the PSSession constructor public. My ideal solution would involve just building the PSSession object with lots of things exposed that we can override at will. Once the object is created, PowerShell just uses it like any other object, but under the hood it handles a custom transport.